framework/base/Application.php

```diff
@@ -456,6 +468,88 @@ public function setVendorPath($path)
+     * This method is provided so that you can configure Dependency Injection Container
+     * when configuring an application.
+     *
+     * @param array $definitions array of definitions. There are two allowed formats of array:
+     * The first format:
+     *  - key: class name, interface name or alias name. The key will be passed to the
+     * [[yii\di\Container::set()]] method as a first argument `$class`.
+     *  - value: the definition associated with `$class`. Possible values are described in
+     * [[yii\di\Container::set()]] documentation for the `$definition` parameter. Will be passed to
+     * the [[yii\di\Container::set()]] method as the second argument `$definition`.
+     *
+     * Example:
+     * ```php
+     * [
+     *     'yii\web\Request' => 'app\components\Request',
+     *     'yii\web\Response' => [
+     *         'class' => 'app\components\Response',
+     *         'format' => 'json'
+     *     ],
+     *     'foo\Bar' => function () {
+     *         $qux = new Qux;
+     *         $foo = new Foo($qux);
+     *         return new Bar($foo);
+     *     }
+     * ]
+     * ```
+     *
+     * The second format:
+     *  - key: class name, interface name or alias name. The key will be passed to the
+     * [[yii\di\Container::set()]] method as a first argument `$class`.
+     * - value: array of two elements. The first element will be passed the [[yii\di\Container::set()]]
+     * method as the second argument `$definition`, the second one - as `$params`.
+     *
+     * Example:
+     * ```php
+     * [
+     *     'foo\Bar' => [
+     *          ['class' => 'app\Bar'],
+     *          [Instance::of('baz')]
+     *      ]
+     * ]
+     * ```
+     *
+     * @see \yii\di\Container::set() to know more about possible values of definitions
+     */
+    public function setContainerDefinitions(array $definitions)
+    {
+        foreach ($definitions as $class => $definition) {
+            if (count($definition) === 2 && array_values($definition) === $definition) {
+                Yii::$container->set($class, $definition[0], $definition[1]);
+            } else {
+                Yii::$container->set($class, $definition);
+            }
+        }
+    }
+
+    /**
+     * Registers a class definitions in [[yii\di\Container|Dependency Injection Container]]
+     * and marks them class as singletons by calling [[yii\di\Container::setSingleton()]].
+     * This method is provided so that you can configure Dependency Injection Container
+     * when configuring an application.
+     *
+     * @param array $singletons array of singleton definitions. See [[setContainerDefinitions]]
+     * for allowed formats of array.
+     *
+     * @see setContainerDefinitions() for allowed formats of $singletons parameter
+     * @see \yii\di\Container::setSingleton() to know more about possible values of definitions
+     */
+    public function setContainerSingletons(array $singletons)
```

Comment:
> setSingletons() is not good as for me, because singletons belong to the Dependency Injection Container, not the Application itself

Same can be said about `aliases`, which are global and do NOT belong to the application.
The orignal issue was justified using `aliases` example. So I can see nothing bad here.

Either we should use single words `singletons` and `definitions` or deny the feature completely saying those do not belong to the application.